Calibration/IsolatedParticles/plugins/IsolatedGenParticles.cc

```diff
@@ -16,55 +16,332 @@
 #include "Geometry/CaloEventSetup/interface/CaloTopologyRecord.h"
 #include "Geometry/CaloTopology/interface/CaloSubdetectorTopology.h"
 #include "Geometry/CaloTopology/interface/HcalTopology.h"
 #include "Geometry/CaloTopology/interface/CaloTopology.h"
-#include "RecoCaloTools/Navigation/interface/CaloNavigator.h"
 
 #include "MagneticField/Engine/interface/MagneticField.h"
 #include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
 
-#include "FWCore/Framework/interface/EventSetup.h"
-#include "FWCore/Framework/interface/ESHandle.h"
+#include "RecoCaloTools/Navigation/interface/CaloNavigator.h"
+
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimGeneral/HepPDTRecord/interface/ParticleDataTable.h"
+#include "SimTracker/Records/interface/TrackAssociatorRecord.h"
+// track associator
+#include "TrackingTools/TrajectoryState/interface/FreeTrajectoryState.h"
+
+// root objects
+#include "TROOT.h"
+#include "TSystem.h"
+#include "TFile.h"
+#include "TH1I.h"
+#include "TH2D.h"
+#include "TProfile.h"
+#include "TDirectory.h"
+#include "TTree.h"
+
+#include <cmath>
 #include <iostream>
 #include <iomanip>
 #include <list>
+#include <memory>
 #include <vector>
-#include <cmath>
 
-#include "Calibration/IsolatedParticles/interface/CaloPropagateTrack.h"
-#include "Calibration/IsolatedParticles/interface/ChargeIsolation.h"
+namespace{
+  class ParticlePtGreater{
+  public:
+    int operator()(const HepMC::GenParticle * p1, 
+		   const HepMC::GenParticle * p2) const{
+      return p1->momentum().perp() > p2->momentum().perp();
+    }
+  };
 
-#include "Calibration/IsolatedParticles/plugins/IsolatedGenParticles.h"
+  class ParticlePGreater{
+  public:
+    int operator()(const HepMC::GenParticle * p1, 
+		   const HepMC::GenParticle * p2) const{
+      return p1->momentum().rho() > p2->momentum().rho();
+    }
+  };
+}
 
-const int IsolatedGenParticles::PBins;
-const int IsolatedGenParticles::EtaBins;
+
+class IsolatedGenParticles : public edm::one::EDAnalyzer<edm::one::SharedResources> {
+
+public:
+  explicit IsolatedGenParticles(const edm::ParameterSet&);
+  ~IsolatedGenParticles() override {}
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void   beginJob() override;
+  void   analyze(const edm::Event&, const edm::EventSetup&) override;
+  void   endJob() override { }
+
+  double deltaPhi(double v1, double v2);
+  double deltaR(double eta1, double phi1, double eta2, double phi2);
+  double deltaR2(double eta1, double phi1, double eta2, double phi2);
+
+  void   fillTrack (GlobalPoint& posVec, math::XYZTLorentzVector& momVec, 
+		    GlobalPoint& posECAL, int pdgId, bool okECAL, bool accpet);
+  void   fillIsolatedTrack(math::XYZTLorentzVector& momVec, 
+			   GlobalPoint& posECAL, int pdgId);
+  void   bookHistograms();
+  void   clearTreeVectors();
+  int    particleCode(int);
+
+  static const int     NPBins_   = 3;
```

Comment:
@bsunanda you could even make them `constexpr`

Reply:
Done